Framework/Algorithms/src/CreateUserDefinedBackground.cpp

```diff
@@ -0,0 +1,244 @@
+    API::ITableWorkspace_sptr &table) const {
+  // Delete blank (zero) rows at the end of the table
+  std::vector<size_t> blankRows;
+  const auto isZero = [](const double n) {
+    return !(fabs(n) > std::numeric_limits<double>::epsilon());
+  };
+  for (size_t i = table->rowCount() - 1; i > 0; i--) {
+    double x, y;
+    API::TableRow row = table->getRow(i);
+    row >> x >> y;
+    if (!isZero(x)) {
+      break;
+    } else if (isZero(y)) {
+      blankRows.push_back(i);
+    }
+  }
+  for (const auto &row : blankRows) {
+    table->removeRow(row);
+  }
+
+  // Sort the table
+  std::vector<std::pair<std::string, bool>> sortArgs;
+  sortArgs.emplace_back(table->getColumn(0)->name(), true);
+  table->sort(sortArgs);
+}
+
+/**
+ * Extend background to limits of data: if it doesn't extend this far already,
+ * insert first and last points of data into it.
+ * @param background :: [input, output] Table of background points to work on
+ * @param data :: [input] Input workspace with data
+ */
+void CreateUserDefinedBackground::extendBackgroundToData(
+    API::ITableWorkspace_sptr &background,
+    const API::MatrixWorkspace_const_sptr &data) const {
+  const auto &xData = data->readX(0);
+  const auto &yData = data->readY(0);
+  const auto xMinMax = std::minmax_element(xData.begin(), xData.end());
+  // If first point > data minimum, insert a new first point
+  if (background->Double(0, 0) > *xMinMax.first) {
+    background->insertRow(0);
+    API::TableRow firstRow = background->getFirstRow();
+    const auto dataPosition = xMinMax.first - xData.begin();
+    firstRow << xData[dataPosition] << yData[dataPosition + 1];
+  }
+  // If last point < data maximum, append a new last point
+  if (background->Double(background->rowCount() - 1, 0) < *xMinMax.second) {
+    API::TableRow lastRow = background->appendRow();
+    const auto dataPosition = xMinMax.second - xData.begin();
+    lastRow << xData[dataPosition] << yData[dataPosition - 1];
+  }
+}
+
+/**
+ * Given a table of background points and the original workspace,
+ * return a new workspace containing interpolated background data.
+ * The same background is assumed for all spectra.
+ * @param background :: [input] Table of background points
+ * @param data :: [input] Original data workspace
+ * @returns :: Workspace containing background to be subtracted
+ */
+API::MatrixWorkspace_sptr
+CreateUserDefinedBackground::createBackgroundWorkspace(
+    const API::ITableWorkspace_const_sptr &background,
+    const API::MatrixWorkspace_const_sptr &data) const {
+  auto outputWS = data->clone();
+
+  const auto &xData = outputWS->readX(0);
+  MantidVec yBackground;
+  MantidVec eBackground(outputWS->blocksize(), 0);
+
+  // Interpolate Y data in the table
+  const auto &lerp = getInterpolator(background, data);
+  const bool isHisto = outputWS->isHistogramData();
+  for (size_t i = 0; i < data->blocksize(); i++) {
+    const double x = isHisto ? (xData[i] + xData[i + 1]) * 0.5 : xData[i];
```

Comment:
Yes, that would be better. I will refactor to take advantage of the new `HistogramData` module.